odoo/models.py

```diff
@@ -6168,9 +6169,21 @@ def filtered(self, func):
             records.filtered("partner_id.is_company")
         """
         if isinstance(func, str):
-            name = func
-            func = lambda rec: any(rec.mapped(name))
-        return self.browse([rec.id for rec in self if func(rec)])
+            def filtered_gen():
+                path = func.split(".")
+                for rec in self:
+                    value = rec
+                    for fname in path:
+                        if not (value := value[fname]):
```

Comment:
I made a mistake when I suggested this. Actually there is a regression in this code: if you call filtered with 'one2many.one_field', it will fail if the one2many contains more than one record (ensure_one). I am very surprised that all tests succeed (could you add a test for this ?).

Also, I propose going back to the old `any(rec.mapped(name))` inside a generator (we can add a shortcut if we want to improve the performance of basic filtering in the future).